ibrcommon/ibrcommon/link/CompatLinkManager.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * NetLinkManager.cpp
+ *
+ * Copyright (C) 2017 IBR, TU Braunschweig
+ *
+ * Written-by: Johannes Morgenroth <jm@m-network.de>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "ibrcommon/config.h"
+#include "CompatLinkManager.h"
+
+#if defined __WIN32__
+#include "ibrcommon/link/Win32LinkManager.h"
+#else
+#if defined HAVE_LIBNL || HAVE_LIBNL2 || HAVE_LIBNL3
+#include "ibrcommon/link/NetLinkManager.h"
+#endif
+#include "ibrcommon/link/PosixLinkManager.h"
+#endif
+
+namespace ibrcommon
+{
+	CompatLinkManager::CompatLinkManager()
```

Comment:
My 2c: I find this to be an excessive layer, there is already LinkManager interface to abstract the implementation. Can everything be wrapped there instead?

Reply:
This was the case before. But now I want to have a singleton in the getInstance() and need to catch the exception of NetLinkManager. Any suggestions?

Reply:
Do I understand correctly:
- the issue is `static NetLinkManager lm;` can throw an exception?
- you don't want this same code in LinkManager due to multithreading?

Reply:
Yes. NetLinkManager can throw an exception.
I know this is not a good design, but this is only an issue on Linux-like systems without netlink (as Windows 10 linux sub-system).

Reply:
Well, bad news is neither current code nor this patch are thread safe when compiled with compliance to C++03 or older. Only C++11 guarantees static members to be initialized properly for this to work with multi threading.

Reply:
Actually, that are good news. If the day comes for IBR-DTN 2.x were we can use C++11, this will be safe. For now it is also safe, because it is called before any thread is created.